crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -162,33 +172,63 @@ fn bidirectional_channel<T>(buffer: usize) -> [BidirectionalChannelHalf<T>; 2] {
     let (tx_first, rx_first) = mpsc::channel(buffer);
     let (tx_second, rx_second) = mpsc::channel(buffer);
     [
-        BidirectionalChannelHalf {
-            tx: tx_first,
-            rx: rx_second,
-        },
-        BidirectionalChannelHalf {
-            tx: tx_second,
-            rx: rx_first,
-        },
+        BidirectionalChannelHalf::new(tx_first, rx_second),
+        BidirectionalChannelHalf::new(tx_second, rx_first),
     ]
 }
 
 struct BidirectionalChannelHalf<T> {
-    tx: mpsc::Sender<T>,
-    rx: mpsc::Receiver<T>,
+    /// Sends messages to the other half
+    tx: mpsc::Sender<Option<T>>,
+    /// Receives messages from the other half
+    rx: mpsc::Receiver<Option<T>>,
+    /// Sends to a background task sending messages to the other half and over MQTT
+    unbounded_tx: mpsc::UnboundedSender<(T, Option<T>)>,
+    /// Used by the background task
+    unbounded_rx: Option<mpsc::UnboundedReceiver<(T, Option<T>)>>,
 }
 
 impl<'a, T> BidirectionalChannelHalf<T> {
-    pub fn send(&'a mut self, item: T) -> futures::sink::Send<'a, mpsc::Sender<T>, T> {
-        self.tx.send(item)
+    fn new(tx: mpsc::Sender<Option<T>>, rx: mpsc::Receiver<Option<T>>) -> Self {
+        let (unbounded_tx, unbounded_rx) = mpsc::unbounded::<(T, Option<T>)>();
+        BidirectionalChannelHalf {
+            tx,
+            rx,
+            unbounded_tx,
+            unbounded_rx: Some(unbounded_rx),
```

Comment:
This feels like a bit of a hack. If `new` instead took the `AsyncClient` and spawned the publisher loop, we wouldn't need this `Option` (and we can remove an argument in the function to spawn the bridge halves).

Reply:
Suggestion applied: https://github.com/thin-edge/thin-edge.io/pull/3122/commits/5783f15acfbf034e51a15af51082c5aca25c8c9c